Allow exporting third-party iOS Frameworks

endragor opened the pull request:
- EditorExportPlugin API now allows adding iOS Frameworks to export.

 - EditorExportPlugin can now distinguish between debug and release
   export.

 - Adds method to DirAccess for recursive copying of a directory.

 - Fixes iOS export to work with Xcode 9 (released recently).

Example usage:

```gdscript
tool
extends EditorPlugin

class SomeFrameworkExportPlugin:
  extends EditorExportPlugin

  func _export_begin(features, debug):
    if debug:
      add_ios_framework("res://SomeDebugFramework.framework")
    add_ios_framework("res://SomeFramework.framework")
    add_ios_plist_content("<key>someKey</key><string>somevalue</string>")

func _init():
  add_export_plugin(FabricExportPlugin.new())
```

endragor committed:
Allow configuring iOS export

 - EditorExportPlugin's _export_begin accepts all the arguments related
   to the current export (is_debug, path, flags).

 - EditorExportPlugin API is extended with methods allowing to configure
   iOS export: add_ios_framework, add_ios_plist_content,
   add_ios_linker_flags, add_ios_bundle_file.

 - iOS export template now contains Godot as a static library so that
   it can be linked with third-party Frameworks and GDNative static
   libraries.

 - Adds method to DirAccess for recursive copying of a directory.

 - Fixes iOS export to work with Xcode 9 (released recently).

reduz commented:
I think debug and release appear as feature tags.. or should..

On Oct 2, 2017 12:19 PM, "Ruslan Mustakov" <notifications@github.com> wrote:

>
>    -
>
>    EditorExportPlugin API now allows adding iOS Frameworks to export.
>    -
>
>    EditorExportPlugin can now distinguish between debug and release
>    export.
>    -
>
>    Adds method to DirAccess for recursive copying of a directory.
>    -
>
>    Fixes iOS export to work with Xcode 9 (released recently).
>
> Example usage:
>
> toolextends EditorPlugin
> class SomeFrameworkExportPlugin:
>   extends EditorExportPlugin
>
>   func _export_begin(features, debug):
>     if debug:
>       add_ios_framework("res://SomeDebugFramework.framework")
>     add_ios_framework("res://SomeFramework.framework")
>     add_ios_plist_content("<key>someKey</key><string>somevalue</string>")
> func _init():
>   add_export_plugin(FabricExportPlugin.new())
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/godotengine/godot/pull/11783
> Commit Summary
>
>    - Allow exporting third-party iOS Frameworks
>
> File Changes
>
>    - *M* core/os/dir_access.cpp
>    <https://github.com/godotengine/godot/pull/11783/files#diff-0> (97)
>    - *M* core/os/dir_access.h
>    <https://github.com/godotengine/godot/pull/11783/files#diff-1> (4)
>    - *M* editor/editor_export.cpp
>    <https://github.com/godotengine/godot/pull/11783/files#diff-2> (68)
>    - *M* editor/editor_export.h
>    <https://github.com/godotengine/godot/pull/11783/files#diff-3> (25)
>    - *M* misc/dist/ios_xcode/export_options.plist
>    <https://github.com/godotengine/godot/pull/11783/files#diff-4> (10)
>    - *M* misc/dist/ios_xcode/godot_ios.xcodeproj/project.pbxproj
>    <https://github.com/godotengine/godot/pull/11783/files#diff-5> (6)
>    - *M* misc/dist/ios_xcode/godot_ios/godot_ios-Info.plist
>    <https://github.com/godotengine/godot/pull/11783/files#diff-6> (1)
>    - *M* platform/iphone/export/export.cpp
>    <https://github.com/godotengine/godot/pull/11783/files#diff-7> (194)
>    - *M* platform/javascript/export/export.cpp
>    <https://github.com/godotengine/godot/pull/11783/files#diff-8> (1)
>    - *M* platform/osx/export/export.cpp
>    <https://github.com/godotengine/godot/pull/11783/files#diff-9> (2)
>
> Patch Links:
>
>    - https://github.com/godotengine/godot/pull/11783.patch
>    - https://github.com/godotengine/godot/pull/11783.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/11783>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z2-TDHsX0mdvrCmDTrbtnVMPE_ZrXks5soP8TgaJpZM4PqzSS>
> .
>

sbeca commented on misc/dist/ios_xcode/export_options.plist:
```diff
@@ -4,7 +4,17 @@
+
+    <key>provisioningProfiles</key>
+    <dict>
+        <key>$identifier</key>Ï
```
Is this weird `Ï` character supposed to be here?

endragor replied:
Yeah, not sure how it got there, removed now, thank you.

reduz commented:
Also, I am not completely sure. Cant the frameworks be added from the ios exported options? I would realy like to avoid having references to iOS from ExportPlugin class.. we dont have references to it from anywhere else in the engine.

endragor commented:
Export options wouldn't allow to make a Godot module or a plugin in Asset Lib that adds additional iOS frameworks. Export options should only be used to provide project-specific configuration, but anything reusable between projects should use a different mechanism, and I think EditorExportPlugin is a good candidate.

If you don't like the higher level API, there is an option to add a single call `void before_export_end(String p_binary_path)`. For iOS this would be called before assembling the project, so the plugin has a chance to modify any files in it. A similar approach is used in Unity, but I don't like it, because it puts a lot of burden on the user. Adding iOS frameworks from a Unity plugin turns into a big mess. UE4, on the other hand, provides high level API that looks clean and is easy to use, similar to my example.

reduz commented:
If Unreal exports this this way and it makes sense to you, feel free to fix and merge.
Also, I think @karroffel changed GDNative to properly export using feature tags, so the iOS and Android code should handle the DLL callback properly (no need to add *.so manually), think you could check with him?

endragor commented:
@reduz @karroffel I've updated this PR with changes related to GDNative and static linking on iOS. Modified `add_shared_object` API as we discussed (added `const Vector<String> &tags` argument). I still think name prefix should be made mandatory and static linking should be exclusive to iOS rather than configurable by user, but I left it for @karroffel to change.

karroffel commented:
untested since I don't have any apple hardware, but it looks good so far!

karroffel commented:
Will need a small rebase after #13093

endragor commented:
Rebased and removed leftover debug output. Also prepared Android exporter changes, but they are based on this branch, so I'll add another PR after this is merged.

haxpor commented:
@endragor Thanks for your effort on pushing things forward for iOS platform. I have questions, hope you can shed some light on.

How should users proceed with creating GDNative that works on iOS? If you have time, could you please check my observation [here](https://github.com/godotengine/godot/issues/15476#issuecomment-355913410) as well, so you will know my context.

In short from above link, there are 2 possible options for users to go
1. Create `.framework` for GDNative and link to Godot project.
   But what's the approach to achieve this without redo the code base for our GDNative which aims for cross-platform. I can produce `.dylib`, but still researching to convert/make it available as `.framework`. Does glue code we have in our library needs to be changed? I'm still researching on this as well.
2. Statically link our GDNative library with Godot.
   Should we use [custom module approach](http://docs.godotengine.org/en/stable/development/cpp/custom_modules_in_cpp.html) via statically linked as seen in Godot 2.1 to Godot 3.x ? Do we need to change code in our GDNative library?

For me, I think 1. is totally should be a way to go as it will respect the flexibility GDNative gives and remains its benefits. But mostly might be about how-to to achieve that.

Thanks in advance for any hints or suggestion you might make.
**PS**: To be clear, I'm in active of trying to find above answers as well.

endragor commented:
@haxpor You should compile your library as static (.a file) and specify it within .gdnlib file and it will be built into iOS bundle during export. Code in your library shouldn't be changed.

haxpor commented:
@endragor Sorry to bother you again. I pulled down 554e95747 , rebuild godot and its export template for iOS. I have the following problem (the same as beta 2). My library code won't change. Tested running on macOS is ok.

![screen shot 2018-01-11 at 4 48 32 pm](https://user-images.githubusercontent.com/673757/34816927-5d349e30-f6f1-11e7-9646-f22629fb3977.png)

aaronfranke commented on editor/editor_export.h:
```diff
@@ -152,6 +164,11 @@ class EditorExportPlatform : public Reference {
 		EditorProgress *ep;
 	};
 
+	struct FeatureContainers {
+		Set<String> features;
+		PoolVector<String> features_pv;
+	};
```
What's the reason for having two structures that hold the same data?
